pkg/multiplexer/headers.go

```diff
@@ -62,3 +63,35 @@ func forwardedForFromRequest(req *http.Request) string {
 
 	return host
 }
+
+type headerResponseWriter struct {
+	headers http.Header
+}
+
+func (hrw *headerResponseWriter) Header() http.Header {
+	return hrw.headers
+}
+
+func (hrw *headerResponseWriter) Write(b []byte) (int, error) {
+	return 0, errors.New("Write not supported on response")
+}
+
+func (rw *headerResponseWriter) WriteHeader(status int) {
+}
+
+var _ http.ResponseWriter = &headerResponseWriter{}
+
+func SetHeaders(results []*Result, rw http.ResponseWriter, r *http.Request, handler http.Handler) {
+	wrapped := &headerResponseWriter{headers: results[0].HeadersWithoutProxyHeaders()}
+	r = r.WithContext(ContextWithResults(r.Context(), results))
```

Comment:
If we went down the custom API route, I think that may also allow us to avoid the need to pass results via context?